src/rust/fs/store/src/lib.rs

```diff
@@ -1461,43 +1476,87 @@ impl Store {
                     f.write_all(bytes).map_err(|e| {
                         format!("Error writing file {}: {:?}", destination.display(), e)
                     })?;
+                    // If the destination existed before we wrote to it, ensure it has the right
+                    // permissions (OpenOptions::mode() only sets permissions on created files).
+                    if let Some(perms) = existing_permissions.as_ref() {
+                        let mut new_perms = perms.clone();
+                        new_perms.set_mode(mode);
+                        std::fs::set_permissions(&destination, new_perms).map_err(|e| {
+                            format!(
+                                "Error setting permissions on file at {}: {e}",
+                                destination.display()
+                            )
+                        })?;
+                    }
                     Ok(())
                 })
                 .await?
             }
         }
     }
 
-    pub async fn materialize_symlink(
-        &self,
-        destination: PathBuf,
-        target: String,
-    ) -> Result<(), StoreError> {
-        // Overwriting a symlink, even with another symlink, fails if it exists. This can occur when
-        // materializing to a fixed directory like dist/. To avoid pessimising the more common case (no
-        // overwrite, e.g. materializing to a temp dir), only remove after noticing a failure.
-        //
-        // NB. #17758, #18849: this is a work-around for inaccurate management of the contents of dist/.
+    // Overwriting a symlink or a hardlink (and even copying, on macos) fails if the target exists.
+    // Since materialization is required to be idempotent (e.g., when materializing over an existing
+    // directory in dist/, or when the sandboxer retries materializing after a restart), we must
+    // handle the case where the target exists by deleting the target.
+    // To avoid pessimising the more common case (no overwrite, e.g. materializing to a temp
+    // dir), only remove after noticing a failure.
+    // NB. #17758, #18849: this is a workaround for inaccurate management of the contents of dist/.
+    async fn _safe_link<'a, 'b, R, F, Fut>(
```

Comment:
This is a generalization of the retry logic previously only applied to symlinks.